kubernetes-itests/src/test/java/io/fabric8/kubernetes/TypedCustomResourceIT.java

```diff
@@ -0,0 +1,239 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.fabric8.kubernetes;
+
+import io.fabric8.commons.AssumingK8sVersionAtLeast;
+import io.fabric8.commons.ClusterEntity;
+import io.fabric8.crd.Pet;
+import io.fabric8.crd.PetSpec;
+import io.fabric8.crd.PetStatus;
+import io.fabric8.kubernetes.api.model.KubernetesResourceList;
+import io.fabric8.kubernetes.api.model.ObjectMetaBuilder;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.WatcherException;
+import io.fabric8.kubernetes.client.dsl.MixedOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import org.arquillian.cube.kubernetes.api.Session;
+import org.arquillian.cube.kubernetes.impl.requirement.RequiresKubernetes;
+import org.arquillian.cube.requirement.ArquillianConditionalRunner;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.Collections;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import static org.awaitility.Awaitility.await;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(ArquillianConditionalRunner.class)
+@RequiresKubernetes
+public class TypedCustomResourceIT {
+  @ArquillianResource
+  KubernetesClient client;
+
+  @ArquillianResource
+  Session session;
+
+  private String currentNamespace;
+
+  private MixedOperation<Pet, KubernetesResourceList<Pet>, Resource<Pet>> petClient;
+
+  @ClassRule
+  public static final AssumingK8sVersionAtLeast assumingK8sVersion = new AssumingK8sVersionAtLeast("1", "16");
+
+  @BeforeClass
+  public static void init() {
+    ClusterEntity.apply(TypedCustomResourceIT.class.getResourceAsStream("/pet-crd.yml"));
```

Comment:
Why don't we declare the CRD using Java instead?

I really think we should push our users to think in Java instead of YAML.
It works against us to mix unnecessary YAMLs, this sends the wrong message. I've seen this in many of the open issues when users post examples of infrastructure defined in YAML resources and then pushed to k8s using our client. "Why are they using YAML in the first place?".
We should really be working our way with our own types and try to send this message: https://www.youtube.com/watch?v=QFn_H_LmMfQ

This point should be true for test suites too unless we're either testing serialization/deserialization, or we're actually testing the given k8s resource and we want to prepare it using YAML first (but then that should use other means of deployment aka kubectl).

For this case we're testing CRs, so defining the CRD using our Java API should be perfectly fine and IMO preferred.